SmartContracts.NET/StrongForce.Core/ForwardAction.cs

```diff
@@ -0,0 +1,40 @@
+using System.Runtime.Serialization;
+
+namespace StrongForce.Core
+{
+	public class ForwardAction : Action
+	{
+		public ForwardAction(Address target, Action nextAction)
+			: base(target)
+		{
+			this.NextAction = nextAction;
+			this.NextAction.ConfigureSender(target);
+		}
+
+		public ForwardAction(Address[] targets, Action finalAction)
+			: base(targets[0])
+		{
+			var wrapped = finalAction;
+
+			// i > 0 is not a typo.
+			for (int i = targets.Length - 1; i > 0; i--)
+			{
+				wrapped.ConfigureSender(targets[i]);
+				wrapped = new ForwardAction(targets[i], wrapped);
+			}
+
+			this.NextAction = wrapped;
+			this.NextAction.ConfigureSender(targets[0]);
+		}
+
+		public Action NextAction { get; }
+
+		public Action FinalAction { get => (this.NextAction as ForwardAction)?.FinalAction ?? this.NextAction; }
```

Comment:
As discussed earlier, we would rarely have more than 10 hops. Currently, this favors correctness over performance, so that the concept can be validated more easily.

Reply:
Okay,  but after the validation we should still make it linear, since it's basically moving this line to the first constructor.